src/tools/tidy/src/deps.rs

```diff
@@ -102,9 +255,63 @@ fn extract_license(line: &str) -> String {
     let first_quote = line.find('"');
     let last_quote = line.rfind('"');
     if let (Some(f), Some(l)) = (first_quote, last_quote) {
-        let license = &line[f + 1 .. l];
+        let license = &line[f + 1..l];
         license.into()
     } else {
         "bad-license-parse".into()
     }
 }
+
+/// Get the dependencies of the crate at the given path using `cargo metadata`.
+fn get_deps(path: &Path, cargo: &Path) -> Resolve {
+    // Run `cargo metadata` to get the set of dependencies
+    let output = Command::new(cargo)
+        .arg("metadata")
+        .arg("--format-version")
+        .arg("1")
+        .arg("--manifest-path")
+        .arg(path.join("Cargo.toml"))
+        .output()
+        .expect("Unable to run `cargo metadata`")
+        .stdout;
+    let output = String::from_utf8_lossy(&output);
+    let output: Output = serde_json::from_str(&output).unwrap();
+
+    output.resolve
+}
+
+/// Checks the dependencies of the given crate from the given cargo metadata to see if they are on
+/// the whitelist. Returns a list of illegal dependencies.
+fn check_crate_whitelist<'a>(
+    whitelist: &'a HashSet<Crate>,
+    resolve: &'a Resolve,
+    krate: Crate<'a>,
+) -> Vec<Crate<'a>> {
+    // Will contain bad deps
+    let mut unapproved = Vec::new();
+
+    // If this dependency is not on the WHITELIST, add to bad set
+    if !whitelist.contains(&krate) {
+        unapproved.push(krate);
+    }
+
+    // Do a DFS in the crate graph (it's a DAG, so we know we have no cycles!)
+    let to_check = resolve
+        .nodes
+        .iter()
+        .find(|n| n.id.starts_with(&krate.id_str()))
+        .expect("crate does not exist");
+
+    for dep in to_check.dependencies.iter() {
+        let krate = Crate::from_str(dep);
+        let mut bad = check_crate_whitelist(whitelist, resolve, krate);
+
+        unapproved.append(&mut bad);
+    }
+
+    // Remove duplicates
```

Comment:
Perhaps this could be passed around as `&mut BTreeSet<Crate<'a>>`?

Reply:
done